lib/render/game.php

```diff
@@ -259,3 +259,43 @@ function RenderRecentGamePlayers($recentPlayerData): void
     echo "</tbody></table>";
     echo "</div>";
 }
+
+function RenderGameProgress(int $numAchievements, int $numEarnedCasual, int $numEarnedHardcore)
+{
+    $pctComplete = 0;
+    $pctHardcore = 0;
+    $pctHardcoreProportion = 0;
+    $title = '';
+
+    if ($numAchievements) {
+        $pctAwardedCasual = ($numEarnedCasual + $numEarnedHardcore) / $numAchievements;
+        $pctAwardedHardcore = $numEarnedHardcore / $numAchievements;
+        $pctAwardedHardcoreProportion = 0;
+        if ($numEarnedHardcore > 0) {
+            $pctAwardedHardcoreProportion = $numEarnedHardcore / ($numEarnedHardcore + $numEarnedCasual);
+        }
+
+        $pctComplete = sprintf("%01.0f", $pctAwardedCasual * 100.0);
+        $pctHardcore = sprintf("%01.0f", $pctAwardedHardcore * 100.0);
+        $pctHardcoreProportion = sprintf("%01.0f", $pctAwardedHardcoreProportion * 100.0);
+
+        if ($numEarnedCasual && $numEarnedHardcore) {
+            $title = " title='$pctHardcore% hardcore'";
+        }
+    }
+
+    echo "<div class='progressbar'>";
+    echo "<div class='completion' style='width:$pctComplete%'$title>";
+    echo "<div class='completionhardcore' style='width:$pctHardcoreProportion%'>";
+    echo "&nbsp;";
+    echo "</div>"; // completionhardcore
+    echo "</div>"; // completion
+
+    if ($pctHardcore >= 100.0) {
+        echo "Mastered<br>";
+    } else {
```

Comment:
Everything looks good to me, if you wanted to add Completed as an alternative display for that scenario of only softcore progress at 100% I believe you can just add something along the lines of this here
```
else if ($pctHardcore == 0.0 && $pctComplete >= 100.0) {
    echo "Completed<br>";
}
```

Reply:
As mentioned above, I think the "100% complete" is fine. If it nags the user into playing hardcore to change the text to Mastered, so be it.